session.go

```diff
@@ -130,30 +138,31 @@ func (c *Client) login(ctx context.Context, password string) error {
 			"trustlabel": []string{c.trustLabel},
 		}
 		if _, err := c.postForm(ctx, EndpointTrust, trustForm); err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	c.session.token = rsp.Token
-	privateKey, err := decryptPrivateKey(rsp.PrivateKeyEncrypted, c.encryptionKey)
+	optPrivateKey, err := decryptPrivateKey(rsp.PrivateKeyEncrypted, encKey)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	c.session.privateKey = privateKey
 
-	return nil
+	return &Session{
+		PasswdIterations: passwdIterations,
+		Token:            rsp.Token,
+		EncryptionKey:    encKey,
+		OptSharingKey:    optPrivateKey,
+	}, nil
 }
 
-func (c *Client) loginHash(password string) string {
-	iterations := c.session.passwdIterations
-	key := encryptionKey(c.user, password, iterations)
-	c.encryptionKey = key
+func loginHashAndEncKey(username string, password string, passwdIterations int) (string, []byte) {
```

Comment:
That's a better function, thank you!
(Previous function was implicitly setting `encryptionKey` on the `Client` which was not great.)